samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java

```diff
@@ -0,0 +1,244 @@
+
+  // Mapping from the output stream to the join spec. Since StreamGraph creates two partial join operators for a join and they
+  // will have the same output stream, this mapping is used to choose one of them as the unique join spec representing this join
+  // (who register first in the map wins).
+  Map<MessageStream, OperatorSpec> outputStreamToJoinSpec = new HashMap<>();
+
+  /**
+   * Returns the JSON representation of a {@link JobGraph}
+   * @param jobGraph {@link JobGraph}
+   * @return JSON of the graph
+   * @throws Exception exception during creating JSON
+   */
+  /* package private */ String toJson(JobGraph jobGraph) throws Exception {
+    JobGraphJson jobGraphJson = new JobGraphJson();
+
+    // build StreamEdge JSON
+    jobGraphJson.streams = new HashMap<>();
+    jobGraph.getSources().forEach(e -> getOrCreateStreamEdgeJson(e, jobGraphJson.streams));
+    jobGraph.getSinks().forEach(e -> getOrCreateStreamEdgeJson(e, jobGraphJson.streams));
+    jobGraph.getIntermediateStreamEdges().forEach(e -> getOrCreateStreamEdgeJson(e, jobGraphJson.streams));
+
+    jobGraphJson.jobs = jobGraph.getJobNodes().stream()
+        .map(jobNode -> buildJobNodeJson(jobNode, jobGraphJson.streams))
+        .collect(Collectors.toList());
+
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    ObjectMapper mapper = new ObjectMapper();
+    mapper.writeValue(out, jobGraphJson);
+    return new String(out.toByteArray());
+  }
+
+  /**
+   * Create JSON POJO for a {@link JobNode}, including the {@link org.apache.samza.operators.StreamGraph} for this job
+   * @param jobNode job node in the {@link JobGraph}
+   * @param streamEdges map of {@link org.apache.samza.execution.JobGraphJsonGenerator.StreamEdgeJson}
+   * @return {@link org.apache.samza.execution.JobGraphJsonGenerator.JobNodeJson}
+   */
+  private JobNodeJson buildJobNodeJson(JobNode jobNode, Map<String, StreamEdgeJson> streamEdges) {
+    JobNodeJson job = new JobNodeJson();
+    job.jobName = jobNode.getJobName();
+    job.jobId = jobNode.getJobId();
+    job.operatorGraph = buildOperatorGraphJson(jobNode);
+    return job;
+  }
+
+  /**
+   * Traverse the {@StreamGraph} and build the operator graph JSON POJO.
+   * @param jobNode job node in the {@link JobGraph}
+   * @return {@link org.apache.samza.execution.JobGraphJsonGenerator.OperatorGraphJson}
+   */
+  private OperatorGraphJson buildOperatorGraphJson(JobNode jobNode) {
+    OperatorGraphJson opGraph = new OperatorGraphJson();
+    opGraph.inputStreams = new ArrayList<>();
+    jobNode.getStreamGraph().getInputStreams().forEach((streamSpec, stream) -> {
+        InputStreamJson inputJson = new InputStreamJson();
+        inputJson.streamId = streamSpec.getId();
+        opGraph.inputStreams.add(inputJson);
+        updateOperatorGraphJson((MessageStreamImpl) stream, inputJson, opGraph);
+      });
+    return opGraph;
+  }
+
+  /**
+   * Traverse the {@StreamGraph} recursively and update the operator graph JSON POJO.
```

Comment:
Minor: Will be good to add more descriptive documentation (and maybe use a more specific name) about what this is doing.